### PR TITLE
Competition Forum - Waiting list and edit events deadline

### PR DESCRIPTION
**Pull request by Jambrose777**
- Added a Waiting list deadline with requirements from WCRP.
- Added Events Change deadline with requirements from WCRP.
- Added additional validation for the registration period to be before the competition.
- Have a "fancy" thing, Whenever the refund date is set and the Waiting list deadline is still blank, it will set the waiting list deadline to the same date. This is in line with what the default for it should be according to @DanielEgdal 
These fields were originally required in the extra registration requirements.

CC: @DanielEgdal on behalf of WCAT
Could you also check out the en.yml file to make sure all the wordings for texts / errors are in line with what we discussed / look good?

image
image

### CODE REVIEW

File: WcaOnRails/config/locales/en.yml [hunk 1]
```diff
@@ -1302,6 +1306,8 @@ en:
       use_stripe_below_html: "The registration fee has to be paid through Stripe below once registered."
       refund_policy_html: "If your registration is cancelled before %{limit_date_and_time} you will be refunded %{refund_policy_percent} of your registration fee."
       no_refunds: "Registration fees won't be refunded under any circumstance."
+      waiting_list_deadline_html: "Registrants on the waiting list may be accepted onto the competitor's list until %{waiting_list_deadline}."
```
Comment (DanielEgdal) [inline WcaOnRails/config/locales/en.yml:1309]: Simply "competitor list" seems better than "competitor's list"

File: WcaOnRails/config/locales/en.yml [hunk 2]
```diff
@@ -507,6 +509,8 @@ en:
         on_the_spot_entry_fee_lowest_denomination: "If this is set to 0, a sentence will be displayed saying that on the spot registrations will be accepted for free."
         refund_policy_percent: "For now this number is informational only, refunds will have to be issued manually. If this is set to 0, a sentence will be displayed saying that registration fees won't be refunded under any circumstance."
         refund_policy_limit_date: "Date after which no more refunds will be issued."
+        waiting_list_deadline_date: "The date when waitlisted registrants will no longer be accepted. If you have a deadline for refunds, this time should not be before this."
```
Comment (DanielEgdal) [inline WcaOnRails/config/locales/en.yml:512]: In the last sentence, WCAT realised that using "this" twice might be a bit confusing. So e.g. "[...] the deadline for being accepted should not be before the deadline for refunds" would be more clear though a bit longer.

File: WcaOnRails/config/locales/en.yml [hunk 3]
```diff
@@ -1447,6 +1453,13 @@ en:
       span_too_many_days: "Competition cannot last more than %{max_days} days."
       invalid_currency_code: "Selected currency code is invalid"
       advancement_condition_must_be_present_for_all_non_final_rounds: "must have an advancement condition to all non final rounds."
+      registration_period_after_start: "Registration period cannot be after start date."
+      waiting_list_deadline_before_registration_close: "Wailting list deadline cannot be before regstrion close date."
```
Comment (DanielEgdal) [inline WcaOnRails/config/locales/en.yml:1457]: Small typo in "registration"

**Review by DanielEgdal — COMMENTED** [review]
Thanks for looking at this! Small comments to some of the wording from when WCAT looked at it.

**Comment by Jambrose777** [comment]
> Thanks for looking at this! Small comments to some of the wording from when WCAT looked at it.

Thanks @DanielEgdal ! I've made those changes, and also added the default text for event changes like we talked about.

@reviewer, we wouldn't want to display the default text for event change deadline for past competitions, as that information would not be true. If you know a way better than how I defined has_event_change_deadline_date? in the competitions model, I would be happy to change implementations.

**Review by jonatanklosko — APPROVED** [review]
Looks good, please adjust this one with master too and it should be ready to go =)

**Comment by jonatanklosko** [comment]
Also there is a conflict in after merging #6065, I guess it's just the last line with migration number in `structure.sql`.

**Comment by jonatanklosko** [comment]
This time there actually are quite some tests failing related to the new validations.

File: WcaOnRails/app/models/competition.rb [hunk 4]
```diff
@@ -213,6 +215,7 @@ def persisted_events_id
   MUST_BE_ANNOUNCED_GTE_THIS_MANY_DAYS = 28
 
   # Time in seconds from 6.2.1 in https://www.worldcubeassociation.org/documents/policies/external/Competition%20Requirements.pdf
+  # 48 hours
   REGISTRATION_OPENING_EARLIEST = 172_800
```
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:219]: For the sake of readability, I think `48.hours.seconds` should work

**Review by gregorbg — COMMENTED** [review]
Just a minor nitpick. You also need to merge to resolve some conflicts before merge.

**Comment by DanielEgdal** [comment]
Is there a chance that this broke all old competitions? I cannot access most competitions from 2018, and all the ones from before return an error. But I can access the pages if I add /edit/admin to their URLs.

**Comment by jonatanklosko** [comment]
The problem is here:

https://github.com/thewca/worldcubeassociation.org/blob/f011af49f11ddbb927232091481a6576e2442996/WcaOnRails/app/models/competition.rb#L918-L920

Old competitions don't have `start_date` set, so it should be checked as elsewhere:

```rb
created_at.present? && created_at > Date.new(2021, 4, 1)
```

I don't have an up to date setup, but hopefully someone can send a quick fix.

**Comment by Jambrose777** [comment]
Im taking care of it

**Comment by gregorbg** [comment]
Thanks for the prompt reaction and fix!

**Pull request merged**